packages/core/src/services/chains/ChainStreamConsumer/StreamConsumer/index.ts

```diff
@@ -0,0 +1,68 @@
+import { CoreTool, ObjectStreamPart, TextStreamPart } from 'ai'
+
+import {
+  ChainEvent,
+  RunErrorCodes,
+  StreamEventTypes,
+  StreamType,
+} from '../../../../constants'
+import { Result } from '../../../../lib'
+import { streamToGenerator } from '../../../../lib/streamToGenerator'
+import { AIReturn, StreamChunk } from '../../../ai'
+import { ChainError } from '../../ChainErrors'
+
+const ERROR_CHUNK_TYPES = ['error', 'length', 'content-filter']
+function withError(chunk: StreamChunk) {
+  return ERROR_CHUNK_TYPES.includes(chunk.type)
+}
+
+export function enqueueChainEvent(
+  controller: ReadableStreamDefaultController,
+  event: ChainEvent,
+) {
+  controller.enqueue(event)
+}
+
+export class StreamConsumer {
+  private controller: ReadableStreamDefaultController
+
+  constructor({ controller }: { controller: ReadableStreamDefaultController }) {
+    this.controller = controller
+  }
+
+  async call(result: Awaited<AIReturn<StreamType>>) {
+    const chunks: StreamChunk[] = []
+    for await (const value of streamToGenerator<
+      TextStreamPart<Record<string, CoreTool>> | ObjectStreamPart<unknown>
+    >(result.data.fullStream)) {
+      chunks.push(value)
+      enqueueChainEvent(this.controller, {
+        event: StreamEventTypes.Provider,
+        data: value,
+      })
+    }
+
+    return this.validate(chunks)
+  }
+
+  /**
+   *  We check that chunk type is not 'error'
+   *  if not we return the first error chunk
+   */
+  private validate(chunks: StreamChunk[]) {
+    const errorChunk = chunks.filter(withError)[0]
+
+    if (!errorChunk) return Result.ok(chunks)
+
+    const error =
+      'error' in errorChunk ? (errorChunk.error as Error) : undefined
+    if (!error) return Result.ok(chunks)
+
+    return Result.error(
+      new ChainError({
+        code: RunErrorCodes.AIRunError,
+        message: error.message,
+      }),
+    )
+  }
+}
```

Comment:
this seems like it could have been a single function instead of a class

Reply:
I can change. But is important?

Reply:
i think using classes when not needed adds needless code

Reply:
But popular demand now this is a function